### PR TITLE
MM-53962: Adding ES8 dependency

### PR DESCRIPTION
**Pull request by agnivade**
**Improvements in the client:**

- Added support for Opensearch and ES v8.
- Completely re-written ES client layer using the official elasticsearch library.
- Much improved logging. Logsettings.Debug will now print out every single elasticsearch request made. Setting ElasticsearchSettings.Trace will print out even more detailed logs.
- Auto retry with backoff implemented on certain HTTP status codes.
- Multi-core bulk indexing. (Although this did not give much improvement as I'd have hoped. The main bottleneck still remains with the DB.)

**Things pending**
- Still some duplication is remaining between the two implementations. I have tried to refactor as much as I could into the `common` package. But the main target is to refactor the tests to a common package so that they don't have to be repeated for each backend.

**Code structure**
- I have divided the code into 2 sub-packages. The parent package is still called elasticsearch for convenience. But later, could be renamed to something else.
```
- elasticsearch
|
| - elasticsearch (ES impl)
| - opensearch (OS impl)
| - common (common stuff)
```

https://mattermost.atlassian.net/browse/MM-53962
https://mattermost.atlassian.net/browse/MM-58503
https://mattermost.atlassian.net/browse/MM-58520

```release-note
- Support for Elasticsearch v8 is added. Beta support for Opensearch is also added.
- A new config setting ElasticsearchSettings.Backend has been added to differentiate between Elasticsearch
or Opensearch. The default value is Elasticsearch. 
- Note that this will break support for AWS Elasticsearch 7.10. The official v8 client only works from Elasticsearch 7.10+ versions. Customers using AWS Elasticsearch are requested to upgrade to AWS Opensearch for future compatibility. Upgrade steps can be found here: https://docs.aws.amazon.com/opensearch-service/latest/developerguide/version-migration.html

Note:
- The value cannot be dynamically changed from the system console while the server is running. One needs to restart
the server if you are switching from one backend to another. And either use mmctl or manually edit the config.
- If you are using Opensearch, you _must_ set the backend to opensearch. Otherwise MM will not work.
- If you are using ES v8, be sure to set `action.destructive_requires_name` to false in elasticsearch.yml to allow for wildcard operations to work.
```

### CODE REVIEW

**Review by mgdelacroix — APPROVED** [review]

**Review by isacikgoz — APPROVED** [review]

**Comment by mickmister** [comment]
> ElasticsearchSettings.Sniff is now deprecated, and enabled by default. There's no way to disable it in the new client.

This setting needs to be removed here as well https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/components/admin_console/elasticsearch_settings.tsx

**Review by isacikgoz — APPROVED** [review]

File: server/public/model/builtin.go [hunk 1]
```diff
@@ -7,3 +7,4 @@ func NewBool(b bool) *bool       { return &b }
 func NewInt(n int) *int          { return &n }
 func NewInt64(n int64) *int64    { return &n }
 func NewString(s string) *string { return &s }
+func NewPointer[T any](t T) *T   { return &t }
```
Comment (agarciamontoro) [inline server/public/model/builtin.go:10]: Neat! Maybe we can remove all the other ones (not in this PR, of course)

**Review by agarciamontoro — APPROVED** [review]
Neat!

**Comment by DHaussermann** [comment]
@agnivade I have done a round of testing on this and it does seem functional when testing opensearch and elasticsearch 8.0. 
- Tested that index can run and succeed
- Tested that indexed data can be retrieved
- Tested across public and private channels as well as DMs and GMs 

I'm not sure if more changes are needed for this specific PR but while exploring this and the indexing I may have found a crash that takes down the server and requires a restart. 

1. 0/5 This seems to happen when I start to rebuild the channel Index (there is no UI feedback for when this job will complete) and then click the _Index Now_ button to bulk index. 

Steps:
1. Enable either provider
2. Start rebuilding channel indexes 
3. Click the _Index Now_ button to bulk index.  
**Observed** indexing job fails and server crashes.

After testing this a bit more it may not be new as it is no specific to opensearch.
Unfortunately I have been unable to see and catch the top of the log  in my terminal. But as far back as I can see the terminal, I have attached the contents in this file. 
[dkh-terminal-logs.txt ](https://github.com/user-attachments/files/16073160/dkh-terminal-logs.txt)

The server log records shows 
`{"timestamp":"2024-07-02 14:40:01.728 -04:00","level":"error","msg":"Unhandled panic in job","caller":"jobs/jobs.go:230","worker_name":"EnterpriseElasticsearchIndexer","job_id":"gtwu9kn87bfoiromej5awpemie","job_type":"elasticsearch_post_indexing","job_create_at":"Jul  2 18:39:47.500","panic":"close of closed channel","job":{"create_at":1719945587500,"data":{"error":"Unable to update the job.","index_channels":"true","index_files":"false","index_posts":"false","index_users":"false","sub_type":"channels_index_rebuild"},"id":"gtwu9kn87bfoiromej5awpemie","last_activity_at":0,"priority":0,"progress":0,"start_at":0,"status":"pending","type":"elasticsearch_post_indexing"},"stack":"goroutine 248 [running]:\nruntime/pprof.writeGoroutineStacks({0x1069c18a8, 0x1400031a600})\n\t/Users/dylanhaussermann/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.21.8.darwin-arm64/src/runtime/pprof/pprof.go:703 +0x6c\nruntime/pprof.writeGoroutine({0x1069c18a8?, 0x1400031a600?}, 0xd0?)\n\t/Users/dylanhaussermann/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.21.8.darwin-arm64/src/runtime/pprof/pprof.go:692 +0x2c\nruntime/pprof.(*Profile).WriteTo(0x1055ee50b?, {0x1069c18a8?, 0x1400031a600?}, 0x102b03554?)\n\t/Users/dylanhaussermann/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.21.8.darwin-arm64/src/runtime/pprof/pprof.go:329 +0x148\ngithub.com/mattermost/mattermost/server/v8/channels/jobs.(*JobServer).HandleJobPanic(0x102afcdd0?, {0x1069ee860, 0x14008aa4dc0}, 0x0?)\n\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-server/server/channels/jobs/jobs.go:229 +0x90\npanic({0x10649e0c0?, 0x1069ac640?})\n\t/Users/dylanhaussermann/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.21.8.darwin-arm64/src/runtime/panic.go:920 +0x26c\ngithub.com/opensearch-project/opensearch-go/v4/opensearchutil.(*bulkIndexer).Close(0x14002196000, {0x1069d6e10, 0x108a2c820})\n\t/Users/dylanhaussermann/go/pkg/mod/github.com/opensearch-project/opensearch-go/v4@v4.0.0/opensearchutil/bulk_indexer.go:221 +0x44\ngithub.com/mattermost/enterprise/elasticsearch/opensearch.(*OpensearchIndexerWorker).DoJob.func1()\n\t/Users/dylanhaussermann/go/src/github.com/mattermost/enterprise/elasticsearch/opensearch/indexing_job.go:270 +0x74\npanic({0x10649e0c0?, 0x1069ac620?})\n\t/Users/dylanhaussermann/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.21.8.darwin-arm64/src/runtime/panic.go:920 +0x26c\ngithub.com/opensearch-project/opensearch-go/v4/opensearchutil.(*bulkIndexer).Add(0x14002196000, {0x1069d6e10, 0x108a2c820}, {{0x1055ea7cf, 0x8}, {0x1055e0852, 0x5}, {0x1400918f660, 0x1a}, 0x0, ...})\n\t/Users/dylanhaussermann/go/pkg/mod/github.com/opensearch-project/opensearch-go/v4@v4.0.0/opensearchutil/bulk_indexer.go:205 +0xd0\ngithub.com/mattermost/enterprise/elasticsearch/opensearch.BulkIndexChan..."}`

Please advise if this should be handled with a different PR

2. In terms of less technical usability concerns, do we need a story to be able to switch between one indexing service to the other from the UI? Or do we need to change the `Elasticsearch` to a more generic term or are these terms still similar enough that this is not needed?

**Comment by agnivade** [comment]
1. I can repro this Dylan. Will look into it.
2. No I don't believe dynamically switching from the UI is really necessary. This is mainly a one time thing that customers will just set once.

**Comment by agnivade** [comment]
@DHaussermann - Please give it another go. Both branches have changes, so you'll have to pull both of them.

**Comment by DHaussermann** [comment]
Tested and passed
- Repeated basic smoke testing outlined above 
- As a precaution, searched data in multiple teams. No changes here. 
- Also regression tested search across files as a precaution
- Tested around Bulk Index and Rebuild channel index jobs
- Jobs queue normally if the user starts one while the other is in progress
- Panic has been resolved

LGTM!

Thanks @agnivade

**Review by DHaussermann — APPROVED** [review]

**Pull request merged**